src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -31,83 +33,334 @@ constexpr float DEFAULT_QUEUE_PRIORITY = 1.0f;
 
 namespace inexor::vulkan_renderer::wrapper {
+namespace {
 
-bool Device::is_extension_supported(const VkPhysicalDevice graphics_card, const std::string &extension_name) {
-    assert(graphics_card);
-    assert(!extension_name.empty());
+/// A function for rating physical devices by type
+/// @param info The physical device info
+/// @return A number from 0 to 2 which rates the physical device (higher is better)
+std::uint32_t device_type_rating(const DeviceInfo &info) {
+    switch (info.type) {
+    case VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU:
+        return 2;
+    case VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU:
+        return 1;
+    default:
+        return 0;
+    }
+}
 
-    std::uint32_t device_extension_count = 0;
+/// Check if a device extension is supported by a physical device
+/// @param extensions The device extensions
+/// @note If extensions is empty, this function returns ``false``
+/// @param extension_name The extension name
+/// @return ``true`` if the required device extension is supported
+bool is_extension_supported(const std::vector<VkExtensionProperties> &extensions, const std::string &extension_name) {
+    return std::find_if(extensions.begin(), extensions.end(), [&](const VkExtensionProperties extension) {
+               return extension.extensionName == extension_name;
+           }) != extensions.end();
+}
 
-    // Query how many device extensions are available.
-    if (const auto result =
-            vkEnumerateDeviceExtensionProperties(graphics_card, nullptr, &device_extension_count, nullptr);
-        result != VK_SUCCESS) {
-        throw VulkanException("Error: vkEnumerateDeviceExtensionProperties failed!", result);
-    }
+/// Transform a ``VkPhysicalDeviceFeatures`` into a ``std::vector<VkBool32>``
+/// @note The size of the vector will be determined by the number of ``VkBool32`` variables in the
+/// ``VkPhysicalDeviceFeatures`` struct
+/// @param features The physical device features
+/// @return A ``std::vector<VkBool32>`` The transformed data
+std::vector<VkBool32> get_device_features(const VkPhysicalDeviceFeatures &features) {
+    std::vector<VkBool32> comparable_features(sizeof(VkPhysicalDeviceFeatures) / sizeof(VkBool32));
+    std::memcpy(comparable_features.data(), &features, sizeof(VkPhysicalDeviceFeatures));
+    return comparable_features;
+}
 
-    if (device_extension_count == 0) {
-        // This is not an error. Some platforms simply don't have any device extensions.
-        spdlog::info("No Vulkan device extensions available!");
-        return false;
-    }
+/// Get a feature description of a ``VkBool32`` value in the ``VkPhysicalDeviceFeatures`` struct by index.
+/// @param index The index of the ``VkBool32`` value in the ``VkPhysicalDeviceFeatures`` struct.
+/// @note If the index is out of bounds, no exception will be thrown, but an empty description will be returned instead.
+/// @return A feature description
+std::string_view get_feature_description(const std::uint32_t index) {
```

Comment:
Would place this function not in the unnamed namespace as it is quite huge. Mabye move it to `vk_tools`?

Reply:
Yeah I thought so as well, but not into the `enumerate` header.

Reply:
Wait we could move that into `representation`. That's exactly where it fits in :)

Reply:
I moved it to `representation` and called it `get_device_feature_description`.